### ISSUE TITLE
Sockets send messages multiple times

### ISSUE THREAD

**Issue by Ghirigoro** [thoov/mock-socket#13]
I'm running into a problem while testing in Node. The server seems to receive more messages than are sent. It gets as many copies of a message as there are sockets connected. So if you have 3 sockets connected the server will receive 3 copies of every message. Below is a Mocha test case I wrote to find the issue in case it helps. 

describe('MockSocket', function() {
		it ('server should only receive as many messages as are sent', function(done) {

			var url = 'ws://localhost:8080';

			var server = new MockServer(url);

			var num_messages_sent = 0;
			var num_messages_received = 0;

			server.on('connection', function (socket) {
				socket.on('message', function (event) {
					console.log("RECEIVED: " + event.data);
					num_messages_received++;
				});
			});

			var s1 = new MockSocket(url);
			s1.onopen = function () {
				num_messages_sent++;
				s1.send("1");
			};

			var s2 = new MockSocket(url);
			s2.onopen = function () {
				num_messages_sent++;
				s2.send("2");
			};

			setInterval(function () {
				assert.equal(num_messages_received, num_messages_sent);
                                done();
			}, 500);
		}
	});

**Comment by thoov** [comment]
@Ghirigoro Thanks for the issue. Let me take a look at it tomorrow and see if I can come up with a fix.

**Comment by thoov** [comment]
@Ghirigoro I believe that I have fixed the issue that you reported. I have pushed an update (v0.1.1) so update and see if it fixes your case.

**Comment by johntimothybailey** [comment]
@thoov There is in an interesting implication with this ticket and the equality checks of functions via `.toString()`

## What I found
I discovered that this bug report and the corresponding test have a potential bug in it that implies there was something broken to begin with. The test actually adds two event listener instances. Each time the 'connection' handler is called (which is twice) it creates a new handler for 'message'
Technically, this is expected in Javascript. While you may not read the code that way, there are actually two functions that get created. This is a fairly common misunderstanding with Runtime execution in Javascript

Here let me show you

The following demonstrates the calls
```
var connectionsCreated  = 0;
  var previousHandler;

  mockServer.on('connection', function(server,socket) {
    connectionsCreated++;
    console.group('Socket Client connected to server');
    console.log("Socket",socket);
    console.log("Server",server);
    console.log("Connections made to server",connectionsCreated);
    var currentHandler = function() {
      console.log("message",arguments);
      numMessagesReceived++;
    };
    server.on('message', currentHandler);
    console.log("Does this one match the previous one",previousHandler === currentHandler); // Notice here they will never equal!
    previousHandler = currentHandler;
    console.groupEnd();
  });
```

### The if you wanted it to only invoke once you need to consider

```
mockSocketA.id = "a";
  mockSocketB.id = "b";

  var serverMessageHandler = function() {
    console.log("message",arguments);
    numMessagesReceived++;
  };

  var connectionsCreated  = 0;
  var previousHandler;

  mockServer.on('connection', function(server,socket) {
    connectionsCreated++;
    console.group('Socket Client connected to server');
    console.log("Socket",socket);
    console.log("Server",server);
    console.log("Connections made to server",connectionsCreated);
    var currentHandler = function() {
      console.log("message",arguments);
      numMessagesReceived++;
    };
    server.on('message', serverMessageHandler);

    console.log("Does this one match the previous one",serverMessageHandler === serverMessageHandler); // Notice here they will always equal!

    previousHandler = currentHandler;
    console.groupEnd();
  });
```
Important line
``console.log("Does this one match the previous one",serverMessageHandler === serverMessageHandler); // Notice here they will always equal!
``

As you can see in the above, I am creating the instance (reference) of the function once and external.

### Result
There is a bug in what was expected. This is really a request  for a feature to stop something unexpected

*The Difference is though that what is expected here is that a function written once will only exist once in Javascript runtime*. This is simply not true. The a function can exist multiple depending on the execution.

# Bottom Line
This report is confusing if we want a method to not be called more than once based on the algorithm that is one thing. Whereas, if we want not a reference of a method to be called based on it's reference that is entirely something else.

@thoov Basically, this prevents the ability for me to use bind() and create unique handlers (context plus the algorithm is what makes a function unique), which is more expected in Javascript than the string representation (what is read by our eyes)

**Comment by johntimothybailey** [comment]
@thoov From what I can tell the fix should be to check **reference equality** instead of the string representation. 

Reference Equality
```
item === listener
```

vs

String Equality
```
item.toString() === listener.toString()
```

Given

```
var item = function(){ console.log("foo") }
var listener = listener.bind(Object.create(null));

item === listener // false
item.toString() === listener.toString() // true
```
